Refactor the export manager logic

jeroenpf opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/8363

## Proposed Changes

This PR refactors the export manager and removes validators. The reason for this is that validators don't validate. They were doing various things that made more sense to be closer to the exporter.

Additionally, this fixes the issue where even when you do an SQL export, it matches the default exporter and exports the zip contents to an SQL file. We now check for the requested file extension as well. If an SQL file is requested, it will match the SQL exporter, which will only export the database.


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
You can run the unit tests and additionally run the following steps:
- Start studio with `export STUDIO_IMPORT_EXPORT=true`
- Go to the import/export tab of a site and try to export the full site and just the SQL file.
- A full site export should create a tar.gz file while the database export should just trigger some console logs
 


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

jeroenpf committed:
Refactor the export manager logic

jeroenpf committed:
Remove unused types

jeroenpf committed:
Fix import order

jeroenpf committed:
Fix linting issue

jeroenpf committed:
Fix rebase issues

fluiddot commented on src/lib/import-export/export/exporters/default-exporter.ts:
```diff
@@ -9,25 +9,56 @@ import { ExportOptions, BackupContents, Exporter } from '../types';
 	}
+	async canHandle(): Promise< boolean > {
+		// Check for supported extension
+		const supported_extension = [ 'tar.gz', 'tzg', 'zip' ].find( ( ext ) =>
```
Nitpick: In JS we rarely use snake case so I'd recommend renaming this to camel case: `supportedExtension`.

katinthehatsite replied:
```suggestion
		const supportedExtension = [ 'tar.gz', 'tzg', 'zip' ].find( ( ext ) =>
```

fluiddot commented on src/lib/import-export/export/exporters/default-exporter.ts:
```diff
@@ -9,25 +9,56 @@ import { ExportOptions, BackupContents, Exporter } from '../types';
-			this.addWpContent( options );
-			await this.addDatabase( options );
+			this.addWpContent( this.options );
+			await this.addDatabase( this.options );
```
Now that `options` is an instance property, I'd remove the argument from these functions and access it directly from the instance.

fluiddot commented on src/lib/import-export/export/exporters/default-exporter.ts:
```diff
@@ -9,25 +9,56 @@ import { ExportOptions, BackupContents, Exporter } from '../types';
+	async export(): Promise< void > {
+		this.backup = await this.getBackupContents();
+		const output = fs.createWriteStream( this.options.backupFile );
+		this.archive = this.createArchive( this.options );
```
Now that `options` is an instance property, I'd remove the argument from these functions and access it directly from the instance.

fluiddot commented on src/lib/import-export/export/exporters/default-exporter.ts:
```diff
@@ -9,25 +9,56 @@ import { ExportOptions, BackupContents, Exporter } from '../types';
+		);
 
-	async export( options: ExportOptions ): Promise< void > {
-		this.emit( ExportEvents.EXPORT_START );
```
I noticed this `emit` call was removed after the refactor. We should revert it to notify when the exportation process start.

fluiddot commented on src/lib/import-export/export/exporters/sql-exporter.ts:
```diff
@@ -1,14 +1,25 @@
+
+	async canHandle(): Promise< boolean > {
+		// Check for extension of the backup file to be sql.
+		if ( ! this.options.backupFile.toLowerCase().endsWith( '.sql' ) ) {
+			return false;
+		}
```
I presume that by default we'll export the database as SQL, but it might be interesting to also allow compressed files. This will be useful for cases when the database is big.

jeroenpf replied:
Thats a good point. For now we decided to export as .SQL but we can change this to always compress it. I can look into this when i implement the actual db export logic.

fluiddot commented on src/components/content-tab-import-export.tsx:
```diff
@@ -50,6 +50,7 @@ export const ExportSite = ( {
 		}
 		const options: ExportOptions = {
 			sitePath: selectedSite.path,
+			siteId: selectedSite.id,
```
I haven't found where this parameter is used when exporting. Do we need it?

jeroenpf replied:
I removed it for now. I added it before i was refactoring the code as I'll likely need it to run the cli commands as part of the db export.

katinthehatsite commented:
>  The reason for this is that validators don't validate. They were doing various things that made more sense to be closer to the exporter.

Can you give a bit more details on this?

katinthehatsite commented on src/lib/import-export/export/exporters/default-exporter.ts:
```diff
@@ -9,25 +9,56 @@ import { ExportOptions, BackupContents, Exporter } from '../types';
 	}
+	async canHandle(): Promise< boolean > {
+		// Check for supported extension
+		const supported_extension = [ 'tar.gz', 'tzg', 'zip' ].find( ( ext ) =>
```
I am wondering if we should move these to `constants.ts` similarly to `ACCEPTED_IMPORT_FILE_TYPES`: https://github.com/Automattic/studio/blob/trunk/src/constants.ts ?

jeroenpf replied:
I am not sure, I think defining constants in one place is useful if they need to be used in different places so perhaps when that becomes necessary we can add them there.

fluiddot reviewed (commented):
> This PR refactors the export manager and removes validators. The reason for this is that validators don't validate. They were doing various things that made more sense to be closer to the exporter.
> 
> Additionally, this fixes the issue where even when you do an SQL export, it matches the default exporter and exports the zip contents to an SQL file. We now check for the requested file extension as well. If an SQL file is requested, it will match the SQL exporter, which will only export the database.

It's true that we use the same validator for all exporters, which is aligned with the approach of unifying the logic within the export logic. However, I presume that the original intent was to provide an open system where we could have exporters for different site structures. Although, I understand the latter might never happen.

I've followed the testing instructions and confirmed it worked as expected. I shared some comments that might be interesting to review before merging the PR. Thanks @jeroenpf for the refactor 🙇 !

jeroenpf commented:
> Can you give a bit more details on this?

Sure:

Previously, our export system was structured with separate components:

- Exporters: Responsible for the actual export process.
- Validators: Determined if a particular exporter could handle a given export request.

The main issues with this approach were:

- Unnecessary Abstraction: The validator for the default exporter checked WordPress site files and directories to determine eligibility. It also grouped files and returned an object, which was only relevant for the default exporter.
- Mismatched Interfaces: The validator interface assumed all exporters needed to handle site files, which wasn't true for the SQL exporter.
- Separation of Concerns: The logic to determine if an exporter could handle a request was separated from the exporter itself.

To address these issues, I refactored the code:

- Removed Validator Classes: I eliminated the separate validator classes entirely.
- Integrated Validation Logic: I moved the 'can handle' logic directly into each exporter class. This allows each exporter to determine its own eligibility for a given export request.
- Simplified Interfaces: By moving the validation logic into the exporters, I eliminated the need for a one-size-fits-all validator interface.

katinthehatsite commented on src/lib/import-export/tests/export/export-manager.test.ts:
```diff
@@ -1,96 +1,63 @@
-		);
-		( path.join as jest.Mock ).mockImplementation( ( ...args ) => args.join( '/' ) );
+		mockExportOptions = {} as ExportOptions;
+		console.log = jest.fn(); // Mock console.log
```
```suggestion
		console.log = jest.fn(); 
```

I am not completely sure the comment is neccessary, it seems pretty self-explanatory. What do you think?

jeroenpf committed:
rename to camel case

jeroenpf committed:
Remove options parameter from methods

jeroenpf committed:
Revert removal of event emit

jeroenpf committed:
Remove site id from export options

jeroenpf committed:
Remove mock commenet

katinthehatsite commented on src/lib/import-export/tests/export/export-manager.test.ts:
```diff
@@ -1,96 +1,63 @@
-import fs from 'fs/promises';
```
what would happen in case no exporter can handle the export? should we cover this in a test case?

jeroenpf replied:
I think this was missing from the original implementation. As of now, nothing will happen but i assume that once the progress bar is implemented, this will be addressed.

fluiddot commented on src/lib/import-export/tests/export/validators/wordpress-export-validator.test.ts:
It would be great if we could cover some of these test cases in the `default-exporter.test.ts` file.

fluiddot reviewed (approved):
LGTM 🎊 !

kozer commented:
> > This PR refactors the export manager and removes validators. The reason for this is that validators don't validate. They were doing various things that made more sense to be closer to the exporter.
> 
> > 
> 
> > Additionally, this fixes the issue where even when you do an SQL export, it matches the default exporter and exports the zip contents to an SQL file. We now check for the requested file extension as well. If an SQL file is requested, it will match the SQL exporter, which will only export the database.
> 
> 
> 
> It's true that we use the same validator for all exporters, which is aligned with the approach of unifying the logic within the export logic. However, I presume that the original intent was to provide an open system where we could have exporters for different site structures. Although, I understand the latter might never happen.
> 
> 
> 
> I've followed the testing instructions and confirmed it worked as expected. I shared some comments that might be interesting to review before merging the PR. Thanks @jeroenpf for the refactor 🙇 !

This was the intention behind it. Indeed. For example we can have different site structures eg: .wpress etc. This will be implemented in the future, ( it was referenced in the original ticket) and this was the reason for a separate validator class.
